### PR TITLE
More detail when expecting expression but encountering bad macro argument

### LINKED ISSUE #71039: Confusing error message in macro expansion: “error: expected expression, found `1 + 1`”
I’ll admit, the whole error message is not too bad if you ignore the first line.

If only it wouldn’t suggest that `1 + 1` is not an expression...

```rust

macro_rules! test {
    (let $p:pat = $e:expr) => {test!(($p,$e))};
    // actual mistake: this should be expr
    //           vvv
    (($p:pat, $e:pat)) => {let $p = $e;};
}

fn foo() {
    test!(let x = 1+1);
}

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=681495576df8424c896258d944c4243b))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: expected expression, found `1 + 1`
  --> src/lib.rs:6:37
   |
6  |     (($p:pat, $e:pat)) => {let $p = $e;};
   |                                     ^^ expected expression
...
10 |     test!(let x = 1+1);
   |     ------------------- in this macro invocation

error: aborting due to previous error

error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

@rustbot modify labels to T-compiler, A-macros, C-enhancement, D-papercut, A-diagnostics

### PR DESCRIPTION
**Pull request by estebank**
On nested macro invocations where the same macro fragment changes fragment type from one to the next, point at the chain of invocations and at the macro fragment definition place, explaining that the change has occurred.

Fix #71039.

```
error: expected expression, found pattern `1 + 1`
  --> $DIR/trace_faulty_macros.rs:49:37
   |
LL |     (let $p:pat = $e:expr) => {test!(($p,$e))};
   |                   -------                -- this is interpreted as expression, but it is expected to be pattern
   |                   |
   |                   this macro fragment matcher is expression
...
LL |     (($p:pat, $e:pat)) => {let $p = $e;};
   |               ------                ^^ expected expression
   |               |
   |               this macro fragment matcher is pattern
...
LL |     test!(let x = 1+1);
   |     ------------------
   |     |             |
   |     |             this is expected to be expression
   |     in this macro invocation
   |
   = note: when forwarding a matched fragment to another macro-by-example, matchers in the second macro will see an opaque AST of the fragment type, not the underlying tokens
   = note: this error originates in the macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? compiler

kinda swamped

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (compiler-errors) [inline tests/ui/macros/trace_faulty_macros.stderr:82]: Specifically, it doesn't explain why `1 + 1` is interpreted as a pattern -- I think the only thing we need to point out is that the non-terminal kind is `pat`

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr:90]: I know, ideally we would actually point at `$e:pat`, but without this context it is even harder to understand that the reason `1 + 1` is understood as a pattern is that there's multiple levels of macro evaluation at play.

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr:82]: I'm not sure how familiar users are with the "non-terminal" terminology.

Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr:82]: I'm not sure how familiar users are with the "non-terminal" terminology.

File: compiler/rustc_parse/src/parser/diagnostics.rs
Comment (fee1-dead) [inline compiler/rustc_parse/src/parser/diagnostics.rs:2171]:
```suggestion
        // in a subsequent macro invocation (#71039).
```

File: tests/ui/macros/trace_faulty_macros.rs
Comment (fee1-dead) [inline tests/ui/macros/trace_faulty_macros.rs:54]: please add an empty newline at the end of this file

**Review by fee1-dead — COMMENTED** [review]
I think it would also be helpful to have a test where an input ident is first accepted as `$x: ident` and then subsequently accepted as an expr then pattern triggering the error. Would be nice to see how the macro backtracking works in that case

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (compiler-errors) [reply tests/ui/macros/trace_faulty_macros.stderr:90]: #71039 is two problems:
1. reinterpreting expr non-terminals as patterns leads to confusing errors with multiple layers of nesting,
2. unrelated to nesting, using pattern non-terminals in expr position itself leads to weird errors:


this is an example of (2.), which would be fixed if we were to point at the `:pat` kind, and which (afaict) isn't helped by this pr?
```
macro_rules! m {
    ($e:pat) => { let x = $e; }
    //~^ ERROR expected expression, found `1`
}

fn main() {
    m!(1);
}
```

Comment (compiler-errors) [reply tests/ui/macros/trace_faulty_macros.stderr:90]: > which (afaict) isn't helped by this pr?

maybe this is a mischaracterization -- the context added by this pr might help a bit, but ideally the only thing we'd need to add a new label for is the `:pat` part.

Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr:90]: Yeah, I think I figured out a way to add the span for `e:pat` as the `def_site` context of the span for `$e`, but I'll need to talk with petrochenkov to see if that's reasonable. If we had that, it could also be used in a bunch of other errors as well.

**Comment by fee1-dead** [comment]
cc #113297

r? compiler

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr]: Changed to do that

File: tests/ui/macros/trace_faulty_macros.stderr
Comment (estebank) [reply tests/ui/macros/trace_faulty_macros.stderr]: Copied the wording from the reference, but would like to expand a little bit on the explanation

**Comment by estebank** [comment]
As expected, the tests affected by the larger type are all macro related, in particular tt-muncher (again, unsurprising). I leave it to the reviewer to determine if such a perf impact is worth it for this tracking. An alternative approach would be to create a new span context for macro arguments and inject those in the macro backtrace as the def_span, but the machinery is not set up to do that easily.

**Review by b-naber — APPROVED** [review]

**Comment by b-naber** [comment]
mh about the regressions: I think this is quite a heavy hit for what this gives us. Can you try to improve perf first?

**Comment by estebank** [comment]
@b-naber I feel like the new perf is much more reasonable.

**Comment by b-naber** [comment]
The remaining perf regressions don't look like noise though. This change does improve diagnostics, but it's more of a "nice-to-have" kind of improvement imo. Ideally we wouldn't have any perf regressions on a PR like this, but I don't have super strong feelings on this. If you feel that this change justifies the perf regressions, then I'm ok with merging this.

**Comment by estebank** [comment]
I'm ok with either removing the context again or eating the cost. The PR originally wasn't pointing at the macro fragments precisely because I assumed there was going to be a perf impact. Leaving it to @compiler-errors to decide which avenue to take.

**Comment by estebank** [comment]
@b-naber given the current perf results, I would like to merge this PR without further changes 😄

**Comment by estebank** [comment]
Ah, I see that you had already approved earlier, contingent on perf @b-naber. If the current code looks reasonable, I'd like to merge this sooner rather than later, as it is a bit bitrotty.

**Pull request merged**
